### ISSUE TITLE
Breakpoints and source files

### ISSUE THREAD

**Issue by mike-lischke** [harold-b/vscode-duktape-debug#53]
While I can now set breakpoints before launching my app and without a wait loop (which was necessary before), I now see a new disturbing behavior. When vscode is launched freshly and I set a breakpoint, everything's going well. The debugger stops at the breakpoint. This works as long no error occured during a run. At some point the debugger starts to open an unrelated source file (usually from the node modules) but quickly switches back to the original source file with the breakpoint. Once an error occurs (or maybe other reasons I haven't spotted) this back switch no longer happens and the debugger stops at a line in that unrelated sourcefile. I can then do a step-in or step-over to end up again at the real source location. The wrong source file is also opened then without an active breakpoint. I'm constantly closing unrelated tabs now after each debugger run.

And as if that wasn't odd enough, after some debugger runs, it also fails to show any variables. That means I can not inspect any values nor is the `VARIABLES` section populated. Switching stack entries in the stacktrace leads to an error saying the right stackframe could not be determined. At this point I have to restart vscode to get back to a usable debugging behavior again.

**Comment by andrewrch** [comment]
Hi Mike thanks for raising this - it's not something I've encountered.

* Is this with or without source maps enabled?
* Are you able to provide a small reproducible example I can have a look at?

**Comment by mike-lischke** [comment]
Thanks for looking at this. Actually, I have a hard time to produce something you could use to verify, because what's running here is a Node.js clone based on duktape, which is running the Jasmine testing framework to execute some tests. Since this is unreleased code I cannot send you the entire project (and it would probably be too much for you to prepare, just to test this). However, I've made this screen cast to show the effect:

![duktape-debugging](https://user-images.githubusercontent.com/2398704/60245051-0532ce00-98bc-11e9-901c-4cedffc3f1b4.gif)

Here I start vscode and run the tests. Nothing is executed actually, all tests are disabled, except for that single empty one you see on startup. What happens is that a file from the node module `colors` is opened (which I use in my testing project). The debug log shows some interesting errors:

```
Attached to duktape debugger.
Protocol ID: 2 20201 external unknown
Finalized Initialization.
removeAllTargetBreakpoints
OUT -> <REQ: 1> <0x17: LISTBREAK>
Status Notification: PAUSE
IN <- <REP: 2> <EOM: 0>
OUT -> <REQ: 1> <0x13: RESUME>
[FE] threadsRequest
IN <- <REP: 2> <EOM: 0>
[FE] stackTraceRequest
OUT -> <REQ: 1> <0x1c: GETCALLSTACK>
[FE] setExceptionBreakPointsRequest
[FE] configurationDoneRequest
IN <- <REP: 2> <str: node_modules/colors/lib/extendStringPrototype.js> <str: > <int: 9> <int: 7> <str: node_modules/colors/lib/extendStringPrototype.js> <str: > <int: 50> <int: 3> <str: undefined> <str: forEach> <int: 0> <int: 0> <str: node_modules/colors/lib/extendStringPrototype.js> <str: > <int: 51> <int: 45> <str: node_modules/colors/lib/index.js> <str: main> <int: 13> <int: 11> <str: undefined> <str: require> <int: 0> <int: 0> <str: lib/coverage.js> <str: main> <int: 33> <int: 11> <str: undefined> <str: require> <int: 0> <int: 0> <str: main.js> <str: global> <int: 30> <int: 15> <str: eval> <str: > <int: 1> <int: 1> <EOM: 0>
OUT -> <REQ: 1> <0x1e: EVAL>
[FE] stackTraceRequest
[FE] threadsRequest
IN <- <REP: 2> <int: 0> <str: [object global]> <EOM: 0>
OUT -> <REQ: 1> <0x1c: GETCALLSTACK>
IN <- <REP: 2> <str: node_modules/jasmine-core/lib/jasmine-core/jasmine.js> <str: > <int: 6589> <int: 25> <str: node_modules/jasmine-core/lib/jasmine-core/jasmine.js> <str: > <int: 1049> <int: 7> <str: node_modules/jasmine-core/lib/jasmine-core/jasmine.js> <str: Spec> <int: 578> <int: 101> <str: node_modules/jasmine-core/lib/jasmine-core/jasmine.js> <str: > <int: 1584> <int: 47> <str: node_modules/jasmine-core/lib/jasmine-core/jasmine.js> <str: > <int: 1613> <int: 21> <str: tests/test-gui/enumeration-specs.js> <str: > <int: 244> <int: 19> <str: node_modules/jasmine-core/lib/jasmine-core/jasmine.js> <str: addSpecsToSuite> <int: 1528> <int: 14> <str: node_modules/jasmine-core/lib/jasmine-core/jasmine.js> <str: > <int: 1493> <int: 30> <str: tests/test-gui/enumeration-specs.js> <str: main> <int: 422> <int: 11> <str: undefined> <str: require> <int: 0> <int: 0> <str: node_modules/jasmine/lib/jasmine.js> <str: > <int: 89> <int: 2> <str: undefined> <str: forEach> <int: 0> <int: 0> <str: node_modules/jasmine/lib/jasmine.js> <str: > <int: 90> <int: 5> <str: node_modules/jasmine/lib/jasmine.js> <str: > <int: 257> <int: 68> <str: main.js> <str: runtTests> <int: 100> <int: 55> <str: main.js> <str: global> <int: 101> <int: 181> <str: eval> <str: > <int: 1> <int: 1> <EOM: 0>
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <ERR: 3> <int: 3> <str: invalid callstack index> <EOM: 0>
Request undefined returned error: Not found : invalid callstack index
OUT -> <REQ: 1> <0x1e: EVAL>
IN <- <REP: 2> <int: 0> <str: Object> <EOM: 0>
IN <- <NFY: 4> <int: 6> <int: 0> <EOM: 0>
Disconnected: Target detached: ( 0 )
[FE] disconnectRequest
```

The `extendStringPrototype.js` file is what is mentioned first in the debug trace and might have to do with the problem as it modifies the built-in String type. FYI: this is a pure JS project, no TS involved here.

**Comment by VladimirButorin** [comment]
Hi, for me debugger also stopped working..  with the same behavior. VS Code

**Comment by VladimirButorin** [comment]
Hi, 
I downgraded my vs code installation to the older version, and debugger starts working again.

Version: 1.26.1
Commit: 493869ee8e8a846b0855873886fc79d480d342de
Date: 2018-08-16T18:38:57.434Z
Electron: 2.0.5
Chrome: 61.0.3163.100
Node.js: 8.9.3
V8: 6.1.534.41
Architecture: x64

**Comment by benjymous** [comment]
Fyi, I don't think this is a VSCode issue - rolling back to 1.26 also rolls back the vscode-debug plugin to the previous version (0.4.x) which is why it starts working.

I think this is actually related to bug #51 - namely the changes for multiple workspace roots added in v0.5

**Comment by andrewrch** [comment]
Hi all, sorry for slow progress on this issue.  I've pushed out a new version to the marketplace which includes a PR:
https://github.com/harold-b/vscode-duktape-debug/pull/54

Although I'm not sure this will fix the issue.  Unfortunately it's not easy for me to verify fixes to this project because everyone has a different use case for duktape and whether the debugger works is heavily dependent on debugger and project configuration.  For instance when you open/compile a source file in duktape you give a name for the file you're compiling.  If the filename string does not match the string that is passed from the VSCode debugger, then basically the debugger just wont work.

My intention is to add more logging over the next few days and push out another update to the marketplace so that we can get some data on what paths are being passed between the debugger and VSCode so that we can make this more robust, and hopefully add some more information to the issues that were open since I pushed out version 0.5.0.

Thanks for your patience and sorry again for slow progress.

**Comment by mike-lischke** [comment]
Unfortunately, debugging is completely broken for me now. vscode starts the debugging session and 2-3 seconds later stops without any information what happened.

**Comment by andrewrch** [comment]
Is that since the update yesterday or generally since the update to 0.5.0. 

I hope with some more logging we'll be able to make some progress.

I've used this version myself with pure JS and it is working for me, so can you paste your vscode launch.json here and also the command you use to run duk. If you embed duk in another project can you give examples of filenames being passed to duktape during source compilation.

**Comment by mike-lischke** [comment]
> Is that since the update yesterday or generally since the update to 0.5.0.
> 
> 

You sent that comment about the upgrade, so I upgraded and tested that.
> I hope with some more logging we'll be able to make some progress.
> 
> I've used this version myself with pure JS and it is working for me, so can you paste your vscode launch.json here and also the command you use to run duk. If you embed duk in another project can you give examples of filenames being passed to duktape during source compilation
> 

Sure:

{
    "version": "2.0.0",
    "configurations": [
        {
            "name": "Debug MGA",
            "type": "duk",
            "request": "attach",
            "address": "localhost",
            "port": 9091,
            "localRoot": "${workspaceRoot}",
            //"preLaunchTask": "tsc",
            "sourceMaps": true,
            "outDir": "${workspaceRoot}/tests/",
            "stopOnEntry": false,
            "debugLog": false
        },
        {
            "name": "Launch & attach MGA",
            "type": "duk",
            "request": "attach",
            "address": "localhost",
            "port": 9091,
            "localRoot": "${workspaceRoot}/",
            "preLaunchTask": "launch-mga",
            "sourceMaps": true,
            "outDir": "${workspaceRoot}/tests",
            "stopOnEntry": false,
            "debugLog": true,
            //"debugServer": 4711,
        }
    ]
}


There are 2 configs: the first one expect that the debugee already runs (which is done by using a separate hotkey to call the `launch-mga` script. The second one does both, launching the test app and attaching the debugger. Nothing gets logged btw. 

Fortunately, it's possible to return to the previous version. After I did that, debugging started work again as before.

**Comment by mike-lischke** [comment]
> Is that since the update yesterday or generally since the update to 0.5.0.
> 
> 

You sent that comment about the upgrade, so I upgraded and tested that.
> I hope with some more logging we'll be able to make some progress.
> 
> I've used this version myself with pure JS and it is working for me, so can you paste your vscode launch.json here and also the command you use to run duk. If you embed duk in another project can you give examples of filenames being passed to duktape during source compilation
> 

Sure:

```json
{
    "version": "2.0.0",
    "configurations": [
        {
            "name": "Debug MGA",
            "type": "duk",
            "request": "attach",
            "address": "localhost",
            "port": 9091,
            "localRoot": "${workspaceRoot}",
            //"preLaunchTask": "tsc",
            "sourceMaps": true,
            "outDir": "${workspaceRoot}/tests/",
            "stopOnEntry": false,
            "debugLog": false
        },
        {
            "name": "Launch & attach MGA",
            "type": "duk",
            "request": "attach",
            "address": "localhost",
            "port": 9091,
            "localRoot": "${workspaceRoot}/",
            "preLaunchTask": "launch-mga",
            "sourceMaps": true,
            "outDir": "${workspaceRoot}/tests",
            "stopOnEntry": false,
            "debugLog": true,
            //"debugServer": 4711,
        }
    ]
}
```

There are 2 configs: the first one expect that the debugee already runs (which is done by using a separate hotkey to call the `launch-mga` script. The second one does both, launching the test app and attaching the debugger. Nothing gets logged btw. 

Fortunately, it's possible to return to the previous version. After I did that, debugging started work again as before.

**Comment by andrewrch** [comment]
Ok thanks. So this is JS but with source maps enabled. I saw your comment before and assumed pure JS with no source maps. 

Fwiw I also have source maps working myself, so it's possible tweaking outDir or sourceRoot might help. It's quite hit and miss though so I hope to be able to make this more robust soon.

> Fortunately, it's possible to return to the previous version. After I did that, debugging started to work again as before.

Previous version 0.5.0? Or 0.4.x?

I will push a new version out tomorrow evening with more path logging which might help narrow down the bug introduced or find a launch.json config which will mitigate the issue.

**Comment by mike-lischke** [comment]
I think source maps don't play a role here. They are just artifacts from a previous attempt to make TS working (which failed, but that's a different story).  I went back to 0.5.0 and debugging works again without any other change.

Tbh. I'm not even sure it's a path that's the problem. It's the debugger that doesn't even start. I don't come to anything that would require a path it seems (but I might be wrong).

**Comment by andrewrch** [comment]
> I think source maps don't play a role here

You were quite right - there was an issue with webpack configuration so the extension published was using the wrong entrypoint.  Fixed and published a new revision (v0.5.2)

**Comment by mike-lischke** [comment]
Yes, thanks, I can confirm that the debugging ability is back with 0.5.4. The issue reported here is still there, however.

**Comment by andrewrch** [comment]
Just to confirm, if source maps are not required, could you please try setting that option to false and observe the behaviour afterwards?

If the debugger is opening apparently random files, I'm assuming this is caused by mapping from a sourcemap somewhere in your node modules.  This seems quite likely because most of the changes that were merged up to the 0.5.0 release were related to source map path resolution.

**Comment by mike-lischke** [comment]
However, that problem occurs since quite a while already - long before I enabled sourcemaps. But I tried now again with disabled source maps and there's no improvement.

Instead I see a new problem: the "Launch & attach MGA" launch config uses a prelaunch task and when I use that the debugger no longer stops at breakpoints. The prelaunch task runs a testee application (with duktape debug support enabled) that is used to attach with your extension. With the first launch config I run this testee application using a hotkey and then only attach to it. With the second config the testee is launched directly by vscode (via task.json). I have no idea how that can cause breakpoints no longer to work.
